### PR TITLE
feat: add data elements to ES index

### PR DESCRIPTION
**Pull request by jczhong84**
- Add below metadata to elastic search index
  - data element names
  - column description
  - data element description
- Add the filter for data elements on search table tab  
image

### CODE REVIEW

File: querybook/webapp/components/DataElement/DataElement.scss [hunk 1]
```diff
@@ -1,3 +1,29 @@
 .DataElementCard {
     max-width: 400px;
 }
+
+.DataElementGroupSelect {
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElement.scss:5]: can you have a separate CSS file for SELECT and GROUP SELECT?

File: querybook/webapp/components/DataElement/DataElementGroupSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import React, { useMemo } from 'react';
+
+import { HoverIconTag } from 'ui/Tag/HoverIconTag';
+
+import { DataElementSelect } from './DataElementSelect';
+
+import './DataElement.scss';
+
+export const DataElementGroupSelect: React.FC<{
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElementGroupSelect.tsx:9]: can you put this under components/DataElementSelect/DataElementGroupSelect?

File: querybook/webapp/components/DataElement/DataElementGroupSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    );
+
+    const handleDataElementSelect = React.useCallback(
+        (dataElement: string) => {
+            updateDataElements([...dataElements, dataElement]);
+        },
+        [dataElements, updateDataElements]
+    );
+
+    const handleTagRemove = React.useCallback(
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElementGroupSelect.tsx:25]: handleDataElementRemove?

File: querybook/webapp/components/DataElement/DataElementSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    miniReactSelectStyles,
+} from 'lib/utils/react-select';
+import { DataElementResource } from 'resource/table';
+import { SimpleReactSelect } from 'ui/SimpleReactSelect/SimpleReactSelect';
+
+import './DataElement.scss';
+
+interface IProps {
+    onSelect: (val: string) => any;
+    existingDataElements?: string[];
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElementSelect.tsx:16]: selectedDataElements

File: querybook/webapp/components/DataElement/DataElementSelect.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+const dataElementReactSelectStyle = makeReactSelectStyle(
+    true,
+    miniReactSelectStyles
+);
+
+export const DataElementSelect: React.FunctionComponent<IProps> = ({
+    onSelect,
+    existingDataElements = [],
+}) => {
+    const [dataElementString, setDataElementString] = React.useState('');
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElementSelect.tsx:28]: dataElementSearchStr, setDataElementSearchStr

File: querybook/webapp/components/DataElement/DataElementSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    return (
+        <div className={'DataElementSelect'}>
+            <SimpleReactSelect
+                creatable={false}
+                value={dataElementString}
+                options={dataElementSuggestions}
+                onChange={(val) => handleSelect(val)}
+                selectProps={{
+                    onInputChange: (newValue) => setDataElementString(newValue),
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElementSelect.tsx:62]: could you do `onInputChange: setDataElementString`

File: querybook/webapp/components/DataElement/DataElementSelect.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                    placeholder: 'filter by data element',
+                    styles: dataElementReactSelectStyle,
+                    noOptionsMessage: () => null,
```
Comment (czgu) [inline querybook/webapp/components/DataElement/DataElementSelect.tsx:65]: noOptionsMessage: NOOP

File: querybook/webapp/components/Search/SearchOverview.tsx [hunk 8]
```diff
@@ -693,7 +711,13 @@ export const SearchOverview: React.FC<ISearchOverviewProps> = ({
                     </div>
                 </div>
                 <div className="search-filter">
-                    <span className="filter-title">Schemas</span>
+                    <span
+                        className="filter-title"
+                        aria-label="Table belongs to ANY selected schemas"
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchOverview.tsx:716]: Table belongs to ONE OF selected schemas

File: querybook/server/logic/data_element.py [hunk 9]
```diff
@@ -24,6 +24,18 @@ def get_data_element_by_name(name: str, session=None):
     return DataElement.get(name=name, session=session)
 
 
+@with_session
+def get_data_elements_by_keyword(keyword, limit=10, session=None):
+    return (
+        session.query(DataElement)
+        .filter(DataElement.name.like("%" + keyword + "%"))
+        .order_by(DataElement.name.asc())
+        .offset(0)
```
Comment (czgu) [inline querybook/server/logic/data_element.py:33]: offset 0 is noop?

File: querybook/server/logic/data_element.py [hunk 10]
```diff
@@ -24,6 +24,18 @@ def get_data_element_by_name(name: str, session=None):
     return DataElement.get(name=name, session=session)
 
 
+@with_session
+def get_data_elements_by_keyword(keyword, limit=10, session=None):
```
Comment (czgu) [inline querybook/server/logic/data_element.py:28]: search_data_elements_by_keyword(keyword: str, ...)

File: querybook/server/logic/elasticsearch.py [hunk 11]
```diff
@@ -538,20 +540,46 @@ def get_table_weight(table_id: int, session=None) -> int:
     return int(math.log2(((num_impressions + num_samples * 10) + 1) + boost_score))
 
 
+def get_unique_data_elements_from_table(table: DataTable) -> list[DataElement]:
+    unique_names = set()
```
Comment (czgu) [inline querybook/server/logic/elasticsearch.py:544]: here is a shorter version:
```py
return {
  data_element.name: data_element
  for data_element in c.data_elements
  for c in table.columns
}.values()
```

File: querybook/server/models/metastore.py [hunk 12]
```diff
@@ -298,6 +298,10 @@ class DataTableColumn(TruncateString("name", "type", "comment"), Base):
         sql.Integer, sql.ForeignKey("data_table.id", ondelete="CASCADE")
     )
 
+    data_elements = relationship(
```
Comment (czgu) [inline querybook/server/models/metastore.py:301]: did you test this would work for both key and value association?

Comment (jczhong84) [reply querybook/server/models/metastore.py:301]: yep, I tried it on a map type.

File: querybook/webapp/components/DataTableTags/utils.ts [hunk 13]
```diff
@@ -26,3 +26,8 @@ export function useRankedTags(tags: ITag[]) {
         [tags]
     );
 }
+
+export function isTagValid(val: string) {
```
Comment (czgu) [inline querybook/webapp/components/DataTableTags/utils.ts:30]: put it under lib/utils/tag.ts?

File: querybook/webapp/components/Search/EntitySelect.tsx [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+interface IEntitySelectProps {
+    selectedEntities: string[];
+    loadEntities: (keyword: string) => ICancelablePromise<{ data: string[] }>;
+
+    showSelected?: boolean;
+    creatable?: boolean;
+    placeholder?: string;
+    onSelect?: (entity: string) => void;
+    onEntitiesChange?: (entities: string[]) => void;
+    isEntityValid?: (entity: string) => boolean;
```
Comment (czgu) [inline querybook/webapp/components/Search/EntitySelect.tsx:23]: validateEntity

**Review by czgu — APPROVED** [review]

**Review by czgu — APPROVED** [review]

**Pull request merged**